### PR TITLE
Fix Windows event handling for window close

### PR DESCRIPTION
**Pull request by metajack**
PostQuitMessage should not be used since there may be multiple windows
open and PostQuitMessage causes the message loop to completely
terminate. Also, after a window is closed, its HWND is gone, and so we
need to use PostThreadMessage to send the wakeup message.

### CODE REVIEW

**Comment by tomaka** [comment]
This change means that the background thread would continue to run forever.

File: src/api/win32/callback.rs [hunk 1]
```diff
@@ -342,7 +332,7 @@ pub unsafe extern "system" fn callback(window: winapi::HWND, msg: winapi::UINT,
 
         x if x == *super::WAKEUP_MSG_ID => {
             use events::Event::Awakened;
-            send_event(window, Awakened);
+            send_event(None, Awakened);
```
Comment (tomaka) [inline src/api/win32/callback.rs]: I'd prefer if this directly loaded the sender from the stash. Having an `Option` in the signature of `send_event` doesn't really make sense.

File: src/api/win32/init.rs [hunk 2]
```diff
@@ -81,8 +81,14 @@ pub fn new_window(window: &WindowAttributes, pf_reqs: &PixelFormatRequirements,
                     break;
                 }
 
-                user32::TranslateMessage(&msg);
-                user32::DispatchMessageW(&msg);   // calls `callback` (see the callback module)
+                // check for thread message to awaken
+                if msg.message == *super::WAKEUP_MSG_ID {
+                    use events::Event::Awakened;
+                    callback::send_event(None, Awakened);
```
Comment (tomaka) [inline src/api/win32/init.rs]: Why is this handled separately instead of being handled by the callback?

Comment (metajack) [inline src/api/win32/init.rs]: callback won't get invoked once there is no window. I tried that first :)

**Comment by metajack** [comment]
Shouldn't there only be one message handling thread servicing multiple Windows? In that case, the thread should survive until termination.

I believe the current design tries to do one thread per window, but this does not seem correct. However, I may be missing something as I'm no Windows expert.

If the one thread per Window is correct, then I'm not sure how we are going to handle wakeups after window close. The wait iterator will will return None (contrary to the documentation saying this isn't possible) because the sender side has been destroyed. In Servo this causes the event loop to spin consuming all the CPU assuming you just loop on None instead of panicking.

File: src/api/win32/callback.rs (hunk 1 above)
Comment (metajack) [inline src/api/win32/callback.rs]: Currently the function verifies the passed in window is the correct one and ignores other messages. Are you suggesting to just remove that filter? That would work too, but I assumed you had it there for a reason.

**Comment by tomaka** [comment]
I don't think there's a reason why we currently need one thread per window, but I know that several win32 functions apply to *the window in the current thread" (for example, changing the cursor when the window is hovered). If one thread is used for multiple windows, we might be blocked.

The situation about that `Awakened` event did not get a lot of thoughts, because I've always considered it as a huge hack and only Servo uses it.

**Pull request closed without merging**